src/Surfnet/StepupMiddleware/ApiBundle/Configuration/EventListener/WhitelistEventListener.php

```diff
@@ -0,0 +1,117 @@
+use Surfnet\StepupMiddleware\CommandHandlingBundle\Configuration\Command\CreateInstitutionConfigurationCommand;
+use Surfnet\StepupMiddleware\CommandHandlingBundle\Pipeline\Pipeline;
+
+final class WhitelistEventListener implements EventListenerInterface
+{
+    /**
+     * @var ConfiguredInstitutionRepository
+     */
+    private $configuredInstitutionRepository;
+
+    /**
+     * @var Pipeline
+     */
+    private $pipeline;
+
+    public function __construct(ConfiguredInstitutionRepository $configuredInstitutionRepository, Pipeline $pipeline)
+    {
+        $this->configuredInstitutionRepository = $configuredInstitutionRepository;
+        $this->pipeline                        = $pipeline;
+    }
+
+    /**
+     * @param DomainMessage $domainMessage
+     */
+    public function handle(DomainMessage $domainMessage)
+    {
+        $event  = $domainMessage->getPayload();
+
+        $classParts = explode('\\', get_class($event));
+        $method = 'apply' . end($classParts);
+
+        if (!method_exists($this, $method)) {
+            return;
+        }
+
+        $this->$method($event, $domainMessage);
+    }
+
+    public function applyWhitelistCreatedEvent(WhitelistCreatedEvent $event)
+    {
+        foreach ($event->whitelistedInstitutions as $whitelistedInstitution) {
+            $institution = new Institution($whitelistedInstitution->getInstitution());
+
+            if ($this->configuredInstitutionRepository->hasConfigurationFor($institution)) {
+                continue;
+            }
+
+            $this->createInstitutionConfigurationFor($institution);
+        }
+    }
+
+    public function applyWhitelistReplacedEvent(WhitelistReplacedEvent $event)
+    {
+        foreach ($event->whitelistedInstitutions as $whitelistedInstitution) {
+            $institution = new Institution($whitelistedInstitution->getInstitution());
+
+            if ($this->configuredInstitutionRepository->hasConfigurationFor($institution)) {
+                continue;
+            }
+
+            $this->createInstitutionConfigurationFor($institution);
+        }
+    }
+
+    public function applyInstitutionsAddedToWhitelistEvent(InstitutionsAddedToWhitelistEvent $event)
+    {
+        foreach ($event->addedInstitutions as $addedInstitution) {
+            $institution = new Institution($addedInstitution->getInstitution());
+
+            if ($this->configuredInstitutionRepository->hasConfigurationFor($institution)) {
+                continue;
+            }
+
+            $this->createInstitutionConfigurationFor($institution);
+        }
+    }
+
+    /**
+     * @param Institution $institution
+     */
+    private function createInstitutionConfigurationFor(Institution $institution)
+    {
+        $command              = new CreateInstitutionConfigurationCommand();
+        $command->UUID        = (string) Uuid::uuid4();
+        $command->institution = $institution;
+
+        $this->pipeline->process($command);
```

Comment:
A dispatcher leaks implementation details, plus it dispatches one specific command in a specific way. 
A factory is indeed creational, it was just a quick example :wink: You might almost want a mediator of some sorts...
Anyway, a decoupling here would be good.